src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -376,6 +376,16 @@ private MsalAccessTokenCacheItem FindAccessTokenCommon
                             .Where(item => item.HomeAccountId.Equals(requestParams.Account?.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase))
                             .ToList();
                 }
+
+                if (tokenCacheItems.Count > 1)
+                {
+                    requestParams.RequestContext.Logger.Info("Filtering by canonical authority...");
+                    tokenCacheItems =
+                        tokenCacheItems
+                            .Where(item => item.Authority.Equals(requestParams.Authority.CanonicalAuthority,
+                                StringComparison.OrdinalIgnoreCase))
+                            .ToList();
+                }
```

Comment:
it kind of makes sense, and then... though I'm thinking which tokens are in the cache what is the token which is not with the CanonicalAuthority?

Reply:
Also not sure if the match should be here or if it should be later, perhaps it should be a after if (count == 1) to evaluate if only the Authorities are different, then we can return the one which has the same Authority as what has been requested? Though this might also be ok... should probably be aligned across libraries though.

---
In reply to: [238930738](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/708#discussion_r238930738) [](ancestors = 238930738)

Reply:
I looked into moving the logic later, after (count==1), but, at that point, we are filtering by EnvironmentAlias and at this point, in this scenario, the envrionmentAlias is the same across tenants, so only one token is returned and we're back in a loop of using the same AT over and over, even though the authority is different.
ex: 
`Silent call authority = [https://login.microsoftonline.com/](https://login.microsoftonline.com/)"72f...ab"/ -> Microsoft
Silent call authority =  [https://login.microsoftonline.com/](https://login.microsoftonline.com/)"913...f7"/ -> iddxb2csamplestesting`

---
In reply to: [238931288](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/708#discussion_r238931288) [](ancestors = 238931288,238930738)

Reply:
ok. thanks for explaining.

---
In reply to: [239285624](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/708#discussion_r239285624) [](ancestors = 239285624,238931288,238930738)